url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/fixed/FixedOps3.java

```diff
@@ -0,0 +1,633 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ejml.alg.fixed;
+
+import org.ejml.data.FixedMatrix3_64F;
+import org.ejml.data.FixedMatrix3x3_64F;
+
+/**
+ * <p>Common matrix operations for fixed sized matrices which are 3 x 3 or 3 element vectors.</p>
+ * <p>DO NOT MODIFY.  Automatically generated code created by GenerateFixedOps</p>
+ *
+ * @author Peter Abeles
+ */
+public class FixedOps3 {
+    /**
+     * <p>Performs the following operation:<br>
+     * <br>
+     * c = a + b <br>
+     * c<sub>ij</sub> = a<sub>ij</sub> + b<sub>ij</sub> <br>
+     * </p>
+     *
+     * <p>
+     * Matrix C can be the same instance as Matrix A and/or B.
+     * </p>
+     *
+     * @param a A Matrix. Not modified.
+     * @param b A Matrix. Not modified.
+     * @param c A Matrix where the results are stored. Modified.
+     */
+    public static void add( FixedMatrix3x3_64F a , FixedMatrix3x3_64F b , FixedMatrix3x3_64F c ) {
+        c.a11 = a.a11 + b.a11;
+        c.a12 = a.a12 + b.a12;
+        c.a13 = a.a13 + b.a13;
+        c.a21 = a.a21 + b.a21;
+        c.a22 = a.a22 + b.a22;
+        c.a23 = a.a23 + b.a23;
+        c.a31 = a.a31 + b.a31;
+        c.a32 = a.a32 + b.a32;
+        c.a33 = a.a33 + b.a33;
+    }
+
+    /**
+     * <p>Performs the following operation:<br>
+     * <br>
+     * a = a + b <br>
+     * a<sub>ij</sub> = a<sub>ij</sub> + b<sub>ij</sub> <br>
+     * </p>
+     *
+     * @param a A Matrix. Modified.
+     * @param b A Matrix. Not modified.
+     */
+    public static void addEquals( FixedMatrix3x3_64F a , FixedMatrix3x3_64F b ) {
+        a.a11 += b.a11;
+        a.a12 += b.a12;
+        a.a13 += b.a13;
+        a.a21 += b.a21;
+        a.a22 += b.a22;
+        a.a23 += b.a23;
+        a.a31 += b.a31;
+        a.a32 += b.a32;
+        a.a33 += b.a33;
+    }
+
+    /**
+     * Performs an in-place transpose.  This algorithm is only efficient for square
+     * matrices.
+     *
+     * @param m The matrix that is to be transposed. Modified.
+     */
+    public static void transpose( FixedMatrix3x3_64F m ) {
+        double tmp;
+        tmp = m.a12; m.a12 = m.a21; m.a21 = tmp;
+        tmp = m.a13; m.a13 = m.a31; m.a31 = tmp;
+        tmp = m.a23; m.a23 = m.a32; m.a32 = tmp;
+    }
+
+    /**
+     * <p>
+     * Transposes matrix 'a' and stores the results in 'b':<br>
+     * <br>
+     * b<sub>ij</sub> = a<sub>ji</sub><br>
+     * where 'b' is the transpose of 'a'.
+     * </p>
+     *
+     * @param input The original matrix.  Not modified.
+     * @param output Where the transpose is stored. If null a new matrix is created. Modified.
+     * @return The transposed matrix.
+     */
+    public static FixedMatrix3x3_64F transpose( FixedMatrix3x3_64F input , FixedMatrix3x3_64F output ) {
+        if( input == null )
+            input = new FixedMatrix3x3_64F();
+
+        output.a11 = input.a11;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;output&#39; (index: 1 | type: FixedMatrix3x3_64F). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;input&#39;. The same parameter type is validated here for example: url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/fixed/BenchmarkInverseFixed.java:75. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).